[R-package] adding routine registration in R package (fixes #1910)

jameslamb opened the pull request:
~**NOTE:** This PR is a work in progress. I branched off of #2901 . The diff for this PR will get smaller once that PR is merged.~

Now that #2901 has been merged, this PR is ready for review!

This fixes this `R CMD CHECK` note:

```
File ‘lightgbm/libs/lib_lightgbm.so’:
  Found no calls to: ‘R_registerRoutines’, ‘R_useDynamicSymbols’

It is good practice to register native routines and to disable symbol
search.

See ‘Writing portable packages’ in the ‘Writing R Extensions’ manual.
```

Changes in this PR:

* Changed `CMakeLists.txt` to not even compile lightgbm_R.cpp in builds not meant for the R package. Non-R users will get slightly faster builds / smaller libraries, and we don't have to annotate different parts of `lightgbm_R.cpp` with `#ifdef LGB_R_BUILD`
* Added calls to `R_registerRoutines` and `R_useDynamicSymbols`

I was very pleasantly surprised to learn that we could address this without changing `LGBM_SE` to `SEXP` everywhere 😀 . Maybe we'll decide to do that in the future to deal with [whatever changes R 4.0.0 introduces](https://cran.r-project.org/doc/manuals/r-devel/NEWS.html).

All the entries in `R_CallMethodDef CallEntries` were created with this Python script. I don't think it should be checked into the repo, as future updates to this should be small, such as those we'd need to do for #2850 .

<details><summary>get-signatures.py (click me)</summary>

```python

IN_FILE = 'include/LightGBM/lightgbm_R.h'
with open(IN_FILE, 'r') as f:
    txt = [
        line.strip() for line in f
    ]

# filter lines
txt = [
    line for line in txt
    if (
        not line.startswith('/')
        and not line.startswith('*')
        and not line.startswith('#')
    )
]

prefix = 'LIGHTGBM_C_EXPORT LGBM_SE '
functions = {}
current_function = None
for line in txt:
    if line.startswith(prefix):
        current_function = line.replace(prefix, '').replace('(', '')
        functions[current_function] = 0
        print(current_function)
    elif line.startswith('LGBM_SE'):
        functions[current_function] += 1
    elif line.startswith(');'):
        current_function = None
    else:
        raise RuntimeError(f"Encountered unknown line: '{line}'")

# create output like
# {"LGBM_BoosterSaveModelToString_R", (DL_FUNC) &LGBM_BoosterSaveModelToString_R, 6},
# {"LGBM_BoosterDumpModel_R", (DL_FUNC) &LGBM_BoosterDumpModel_R, 6}
longest_func_length = max([
    len(function_name)
    for function_name in functions.keys()
])
print(f"The longest function is {longest_func_length} characters")
for func_name, num_args in functions.items():
    out = '{"'
    out += func_name
    out += '"'
    out += " " * (longest_func_length - len(func_name))
    out += ', (DL_FUNC) &'
    out += func_name
    out += " " * (longest_func_length - len(func_name))
    out += ', '
    out += str(num_args)
    out += '},'
    print(out)
```

</details>

jameslamb commented:
noticing a weird there where whenever I don't have a `build_script` block, AppVeyor does this: 

![image](https://user-images.githubusercontent.com/7608904/77255594-ac1f7d00-6c36-11ea-98ba-ce064cb23d74.png)

StrikerRUS commented:
@jameslamb 
> noticing a weird there where whenever I don't have a build_script block, AppVeyor does this:

To disable this you should have `build: false` at the top of your `appveyor.yml` file like here
https://github.com/RGF-team/rgf/blob/ba62e1f3bbba0b1c3e7b64b6e1aef5c5892e3b42/.appveyor.yml#L1

> This causes AppVeyor to look for a Visual Studio project or solution file in the root directory of your project, and use that to do the build. The behavior of an MSBuild mode build can be controlled by the build: entry in appveyor.yml.
https://www.appveyor.com/docs/build-configuration/#build-mode

jameslamb commented:
> @jameslamb
> 
> > noticing a weird there where whenever I don't have a build_script block, AppVeyor does this:
> 
> To disable this you should have `build: false` at the top of your `appveyor.yml` file like here
> https://github.com/RGF-team/rgf/blob/ba62e1f3bbba0b1c3e7b64b6e1aef5c5892e3b42/.appveyor.yml#L1
> 
> > This causes AppVeyor to look for a Visual Studio project or solution file in the root directory of your project, and use that to do the build. The behavior of an MSBuild mode build can be controlled by the build: entry in appveyor.yml.
> > https://www.appveyor.com/docs/build-configuration/#build-mode

hey thank you! And sorry, I didn't mean to post this on this PR and notify you. Meant to leave it as a note for myself on another PR I'm working on in my fork. Too many tabs open 😬

jameslamb commented:
This PR has been rebased to `master`

jameslamb commented:
> This PR has been rebased to `master`

With these changes plus the other changes merged in over the last few days, `Rscript build_r.R` produces a package with 0 ERRORS, 0 WARNINGS, and 0 NOTES* on macOS (with R 3.6.1) and two different versions of Ubuntu!

<details><summary>(0 / 0 / 0) my mac (`gcc`, macOS 10.14)</summary>

```
* using log directory ‘/Users/jlamb/repos/LightGBM/lightgbm.Rcheck’
* using R version 3.6.1 (2019-07-05)
* using platform: x86_64-apple-darwin15.6.0 (64-bit)
* using session charset: UTF-8
* using option ‘--as-cran’
* checking for file ‘lightgbm/DESCRIPTION’ ... OK
* checking extension type ... Package
* this is package ‘lightgbm’ version ‘2.3.2’
* package encoding: UTF-8
* checking CRAN incoming feasibility ... NOTE
Maintainer: ‘Guolin Ke <guolin.ke@microsoft.com>’

New submission

License components with restrictions and base license permitting such:
  MIT + file LICENSE
File 'LICENSE':
  The MIT License (MIT)

  Copyright (c) Microsoft Corporation

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.

The Date field is over a month old.
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for executable files ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking for sufficient/correct file permissions ... OK
* checking serialization versions ... OK
* checking whether package ‘lightgbm’ can be installed ... OK
* checking installed package size ... OK
* checking package directory ... OK
* checking for future file timestamps ... OK
* checking DESCRIPTION meta-information ... OK
* checking top-level files ... OK
* checking for left-over files ... OK
* checking index information ... OK
* checking package subdirectories ... OK
* checking R files for non-ASCII characters ... OK
* checking R files for syntax errors ... OK
* checking whether the package can be loaded ... OK
* checking whether the package can be loaded with stated dependencies ... OK
* checking whether the package can be unloaded cleanly ... OK
* checking whether the namespace can be loaded with stated dependencies ... OK
* checking whether the namespace can be unloaded cleanly ... OK
* checking use of S3 registration ... OK
* checking dependencies in R code ... OK
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... OK
* checking Rd files ... OK
* checking Rd metadata ... OK
* checking Rd line widths ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... OK
* checking Rd contents ... OK
* checking for unstated dependencies in examples ... OK
* checking contents of ‘data’ directory ... OK
* checking data for non-ASCII characters ... OK
* checking data for ASCII and uncompressed saves ... OK
* checking line endings in C/C++/Fortran sources/headers ... OK
* checking line endings in Makefiles ... OK
* checking compilation flags in Makevars ... OK
* checking for GNU extensions in Makefiles ... OK
* checking for portable use of $(BLAS_LIBS) and $(LAPACK_LIBS) ... OK
* checking use of SHLIB_OPENMP_*FLAGS in Makefiles ... OK
* checking pragmas in C/C++ headers and code ... OK
* checking compilation flags used ... OK
* checking compiled code ... OK
* checking examples ... OK
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ...
  Running ‘testthat.R’ [16s/14s]
 OK
* checking PDF version of manual ... OK
* DONE

Status: 1 NOTE
```

</details>

(0/0/0) Azure DevOps linux (`gcc`, Ubuntu 14.04) ([link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=5593&view=logs&j=a0acac97-df52-5651-dae8-9fb37a4728fc&t=d2c32b51-981b-5bcd-fe31-029dd8d8998c))
(0/0/0) Azure DevOps macOS (`clang`, macOS 10.14) ([link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=5593&view=logs&j=2e42d92c-6989-5c5a-9adb-4d8d43874989&t=7c5d831f-c13d-5c19-c317-c20b1d55f3d3))
(0/0/0) Travis linux (`clang`, Ubuntu 18.04) ([link](https://travis-ci.org/github/microsoft/LightGBM/jobs/666137401?utm_medium=notification&utm_source=github_status))
(0/0/0) Travis macOS (`gcc`, macOS 10.14) ([link](https://travis-ci.org/github/microsoft/LightGBM/jobs/666137408?utm_medium=notification&utm_source=github_status))

*Some of these logs have 1 or two NOTEs, but they're not actually issues that will show up on CRAN. If you test a package that doesn't exist on CRAN today, you get a "hey this is a new package  or new maintainer" note. The other one I see is about some `Suggests` dependencies not being available at testing time. Those won't be an issue on CRAN either because CRAN will install all your `Suggests` dependencies if they come from CRAN (which ours do)*

StrikerRUS commented:
This is awesome! 🎉 

Also, I remember there are some NOTEs about the latest release date:
- the same version is available on CRAN
   > Insufficient package version (submitted: 1.0.5, existing: 1.0.5)
-  new release was just uploaded 
- the latest release is more than one month old
   > The Date field is over a month old.
    https://travis-ci.org/github/RGF-team/rgf/jobs/665940654#L3000

Maybe we should start counting NOTEs at CI and allow only 2?

jameslamb commented:
> Maybe we should start counting NOTEs at CI and allow only 2?

Good point about the other NOTEs, we may encounter those at some point. I agree, we should start to only allow 2 so we don't have things like #2928 pop back up.

From what I've seen so far on #2936  we have two other NOTEs that are Windows-specific that I'll fix in a followup PR (https://github.com/microsoft/LightGBM/pull/2936#issue-392124191). I'm going to add these and some other comments on the current state to #629 since we're tracking progress there.

StrikerRUS commented on CMakeLists.txt:
```diff
@@ -235,7 +235,14 @@ file(GLOB SOURCES
+  add_library(_lightgbm SHARED src/c_api.cpp src/lightgbm_R.cpp ${SOURCES})
+else()
+  add_library(_lightgbm SHARED src/c_api.cpp ${SOURCES})
+endif(BUILD_FOR_R)
```
I think the following will a little bit clearer.

```suggestion
add_library(_lightgbm SHARED src/c_api.cpp ${SOURCES})
# Only build the R part of the library if building for
# use with the R package
if(BUILD_FOR_R)
  add_library(_lightgbm SHARED src/lightgbm_R.cpp ${SOURCES})
endif(BUILD_FOR_R)
```

Also, I believe we can link R libraries only for  `_lightgbm` target, because R package doesn't use executable file.
https://github.com/microsoft/LightGBM/blob/0341906c45666ccaeca92ee5fbee17fa22282470/CMakeLists.txt#L310-L313

jameslamb replied:
oh I didn't realize you can build it up like that. That's definitely clearer, I agree. And agree with the suggestion to only link into `_lightgbm`. I just tried that locally and it worked well.

StrikerRUS replied:
Seems I was wrong and `add_library` can be used only once for the same target. Instead, `target_sources` should be used, but it's available only in >=3.1 version. So, I think we can create a list of additional sources (refer to `SOURCES`) conditionally before executing `add_library`.

jameslamb replied:
ok let me try that

jameslamb replied:
tried in https://github.com/microsoft/LightGBM/pull/2911/commits/34f4b0c2ca1970731f7b7df9e1ee28ea6dba3039

jameslamb commented:
> Should we exclude `lightgbm_R.cpp` from [VS solution](https://github.com/microsoft/LightGBM/tree/master/windows)?

@StrikerRUS I think we should. I've been experimenting with MSBuild on my fork, trying to add something like [the approach in the Python package](https://github.com/microsoft/LightGBM/blob/master/python-package/setup.py#L135-L148) but struggling with it. I think we should not try to build the R package with MSBuild (just sticking to MinGW or MSVC bundled with Visual Studio).

I'll remove it from those files.

guolinke commented:
@jameslamb 
I guess the `MSBuild` you refer to is the `LightGBM.sln` (VS solution) ?

jameslamb commented:
> @jameslamb
> I guess the `MSBuild` you refer to is the `LightGBM.sln` (VS solution) ?

Yeah that's right. Like this approach in the Python package: https://github.com/microsoft/LightGBM/blob/master/python-package/setup.py#L135-L148

Sorry, not that familiar with `MSBuild` and Visual Studio so I might not be referring to those precisely.

StrikerRUS commented on CMakeLists.txt:
```diff
@@ -235,7 +235,15 @@ file(GLOB SOURCES
 
 add_executable(lightgbm src/main.cpp ${SOURCES})
-add_library(_lightgbm SHARED src/c_api.cpp src/lightgbm_R.cpp ${SOURCES})
+set(SOURCES "${SOURCES} src/c_api.cpp")
```
I think `append` should fix CI failures
https://cmake.org/cmake/help/v3.12/command/list.html#append

jameslamb replied:
ah ok, I shouldn't have assumed it was just a space-delimited list. Let me try that

jameslamb replied:
fixed in https://github.com/microsoft/LightGBM/pull/2911/commits/8d5b652c43383e66ca0710498c10eb5b915f9180

jameslamb commented:
> > Maybe we should start counting NOTEs at CI and allow only 2?
> 
> Good point about the other NOTEs, we may encounter those at some point. I agree, we should start to only allow 2 so we don't have things like #2928 pop back up.
> 
> From what I've seen so far on #2936 we have two other NOTEs that are Windows-specific that I'll fix in a followup PR ([#2936 (comment)](https://github.com/microsoft/LightGBM/pull/2936#issue-392124191)). I'm going to add these and some other comments on the current state to #629 since we're tracking progress there.

just rebased to `master` to pick up the changes from #2942 and changed the allowed number of notes from 3 to 2

StrikerRUS commented:
@jameslamb 
> And agree with the suggestion to only link into `_lightgbm`.

Seems that corresponding commit was lost.

jameslamb commented:
> @jameslamb
> 
> > And agree with the suggestion to only link into `_lightgbm`.
> 
> Seems that corresponding commit was lost.

ah! will fix it

jameslamb commented:
> LGTM for CMake and VS parts!

Thanks! I won't merge until @Laurae2 or @guolinke take a look

Laurae2 commented on R-package/src/install.libs.R:
```diff
@@ -163,12 +163,21 @@ if (!use_precompile) {
+  if (file.exists(symbols_file)) {
+    file.copy(symbols_file, dest, overwrite = TRUE)
+  }
+
```
You can fix line breaks on another PR (just stylistic).

guolinke commented:
@Laurae2 sorry, I cannot test it now. I am currently WFH on a remote machine, and the VS require a restart:
![image](https://user-images.githubusercontent.com/16040950/78096617-a97a1000-740c-11ea-92b6-b86742ea5d45.png)

But I cannot restart it, otherwise, it will be stuck in BitLocker, which I cannot enter password remotely 😢

guolinke commented:
@jameslamb Is that possible to test the windows installation over CI (azure pipeline) ?

jameslamb commented:
> @jameslamb Is that possible to test the windows installation over CI (azure pipeline) ?

I can do it on CI! I'll do that right now

jameslamb commented:
> > @jameslamb Is that possible to test the windows installation over CI (azure pipeline) ?
> 
> I can do it on CI! I'll do that right now

@guolinke I created a branch on my fork off of #2936. You can see the PR at https://github.com/jameslamb/LightGBM/pull/12/files

This worked for MinGW ([AppVeyor](https://ci.appveyor.com/project/jameslamb/lightgbm/builds/31868673)) and MSVC ([Azure](https://dev.azure.com/JayLamb20/jameslamb-lightgbm/_build/results?buildId=248&view=logs&j=87d556e4-1858-51ea-bb1c-8e61473c09c8&t=71de8ec8-f901-5587-9a05-3e273fd9dd7e)). You can see in the logs that installation worked, all the tests worked, and this `R CMD check` note is gone:

```
File ‘lightgbm/libs/lib_lightgbm.so’:
  Found no calls to: ‘R_registerRoutines’, ‘R_useDynamicSymbols’
```

StrikerRUS commented on R-package/src/install.libs.R:
```diff
@@ -163,12 +163,21 @@ if (!use_precompile) {
   print(paste0("Found library file: ", src, " to move to ", dest))
   file.copy(src, dest, overwrite = TRUE)
+
+  symbols_file <- file.path(R_PACKAGE_SOURCE, "src", "symbols.rds")
```
I think usage of `source_dir` here will simplify the code
https://github.com/microsoft/LightGBM/blob/6ba84d540c0132d4a603366fc91d4ccaa7b46d69/R-package/src/install.libs.R#L29

jameslamb replied:
Good point! I fixed a few cases of that here: https://github.com/microsoft/LightGBM/commit/1666f67859d2b0635c0b1fd45a26493319d8fbce

I also rebased to `master` to get recent changes

jameslamb commented:
Thanks for the reviews!

Going to merge since CI checks are passing and I think I've addressed all remaining comments. Now that this is merged, everyone building `LightGBM`  in settings other than the R package is going to get faster builds and a smaller `lib_lightgbm` 😀
